compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -5188,110 +5317,45 @@ private boolean isFunctionPointer(BSymbol funcSymbol) {
                 && (funcSymbol.flags & Flags.NATIVE) != Flags.NATIVE;
     }
 
-    private void reportErrorDetailMissmatchError(BLangErrorConstructorExpr errorConstructorExpr,
-                                                 BRecordType targetErrorDetailRec,
-                                                 BRecordType recordType) {
-        boolean detailedErrorReported = false;
-        Set<String> checkedFieldNames = new HashSet<>();
-        for (Map.Entry<String, BField> fieldEntry : targetErrorDetailRec.fields.entrySet()) {
-            checkedFieldNames.add(fieldEntry.getKey());
-            BField argField = recordType.fields.get(fieldEntry.getKey());
-            if (argField == null && !Symbols.isOptional(fieldEntry.getValue().symbol)) {
-                dlog.error(errorConstructorExpr.pos, DiagnosticErrorCode.MISSING_ERROR_DETAIL_ARG, fieldEntry.getKey());
-                detailedErrorReported = true;
-            } else if (!types.isAssignable(argField.type, fieldEntry.getValue().type)) {
-                dlog.error(errorConstructorExpr.pos, DiagnosticErrorCode.INVALID_ERROR_DETAIL_ARG_TYPE,
-                        fieldEntry.getKey(), fieldEntry.getValue().type, argField.type);
-            }
-        }
+    private List<BLangNamedArgsExpression> checkProvidedErrorDetails(BLangErrorConstructorExpr errorConstructorExpr,
+                                                                     BType expectedType) {
+        List<BLangNamedArgsExpression> namedArgs = new ArrayList<>();
+        for (BLangNamedArgsExpression namedArgsExpression : errorConstructorExpr.namedArgs) {
+            BType target = getErrorCtorNamedArgTargetType(namedArgsExpression, expectedType);
 
-        for (Map.Entry<String, BField> fieldEntry : recordType.fields.entrySet()) {
-            if (!checkedFieldNames.contains(fieldEntry.getKey())) {
-                BField field = fieldEntry.getValue();
-                if (targetErrorDetailRec.sealed) {
-                    dlog.error(errorConstructorExpr.pos,
-                            DiagnosticErrorCode.UNKNOWN_DETAIL_ARG_TO_CLOSED_ERROR_DETAIL_REC,
-                            fieldEntry.getKey(), targetErrorDetailRec);
-                    detailedErrorReported = true;
-                } else if (!types.isAssignable(field.type, targetErrorDetailRec.restFieldType)) {
-                    dlog.error(errorConstructorExpr.pos, DiagnosticErrorCode.INVALID_ERROR_DETAIL_REST_ARG_TYPE,
-                            fieldEntry.getKey(), targetErrorDetailRec);
-                    detailedErrorReported = true;
-                }
+            BLangNamedArgsExpression clone = nodeCloner.clone(namedArgsExpression);
```

Comment:
Any particular reason to use a clone here? This is called only once the exact type is found, or if there is no matching type, right?

Reply:
I want to do L:5329 and L:5331 conditionally. `checkProvidedErrorDetails` is called for both success case and failure case.

Reply:
Isn't it only for the success or failure case though, rather than the compatibility check with unions? So can't we directly type check it? Or we need to call checkExpr twice, right?

Reply:
What I want is to use `checkExpr` with a target type (contextually expected type), when I use it and there is an error, something like an expected detail arg type is `string` but what's provided is an `int` literal, using the `checkExpr` with target type generates an extra error message which is wrong as well

https://github.com/rdhananjaya/ballerina-lang/blob/3946651f9686595b9f99fd4e98cfeffa1c603f62/tests/jballerina-unit-test/src/test/resources/test-src/expressions/errorconstructorexpr/error-constructor-expr-negative.bal#L37

If I just use the `checkExpr` with target type I get this extra diagnostic ```invalid arg type in error detail field 'c', expected 'string', found 'other'```

Reply:
For the positive case, wouldn't this always be an additional check? There will be one and only one valid error type, right?

For the negative case can't we merge this with L2760, and iterate by named args and 
1. if the detail type has a field of the particular type - use that to type check the expression
2. if the detail type is closed and it is an additional field log the relevant error
3. if the detail type is open with individual fields and this is an additional field log the relevant error
4. if a required arg is not provided, log the relevant error?